os_migrate/plugins/module_utils/resource.py

```diff
@@ -0,0 +1,236 @@
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import ansible_collections.os_migrate.os_migrate.plugins.module_utils \
+    as collection_utils
+
+class Resource():
+
+    resource_type = 'UNDEFINED'
+
+    # Took the idea of property lists for serialization from Ryan's
+    # code, will mark the patch co-authored.
+
+    # Serialization
+    params_from_sdk = []
+    params_from_ref = []
+    info_from_sdk = []
+    info_from_refs = []
+
+    # Seserialization
+    sdk_params_from_params = []
+    sdk_params_from_ref = []
+
+    # This method should only be called by when creating subclasses
+    # and Resource should not be instantiated directly, not sure if we
+    # have a way to enforce that in Python.
+    def __init__():
+        self.data = {
+            const.RES_TYPE: None,
+            const.RES_PARAMS: {},
+            const.RES_INFO: {},
+        }
+
+    # This constructor will not have to be overriden by child classes,
+    # will work OOTB.
+    @classmethod
+    def from_data(cls, data):
+        res_type = data.get('type', None)
+        if res_type != self.resource_type:
+            raise UnexpectedResourceType(self.resource_type, res_type)
+
+        obj = cls()
+        # Barring resource type errors, we allow Resource subclass
+        # construction with invalid data so that we don't break on
+        # loading a data file, but we explicitly fail with full list
+        # of errors (rather than a single exception) when all
+        # resources are validated with is_data_valid() and/or
+        # data_errors().
```

Comment:
This "load anything, validate explicitly" vs. "validate always on load" is something that could be challenged though. There are different benefits and drawbacks to each.